### PR TITLE
replace `#[async_trait]` with native async trait for trait `Resolver`

### PR DESCRIPTION
**Pull request by AbnerZheng**
Refer to: https://github.com/paradigmxyz/reth/issues/6657

The `native async trait` in Rust has a limitation where it requires explicit declaration if it implements the Send or non-Send version of a trait.

For instance, consider the following trait:
```rust
trait HttpService {
    async fn fetch(&self, url: Url) -> HtmlBody;
}
```
This desugars to:
```rust
trait HttpService {
    fn fetch(&self, url: Url) -> impl Future<Output = HtmlBody>;
}
```
By default, the returned Future does not have the Send bound, which is necessary for a multithreaded runtime. Without using `trait_variant` to create a multithreaded version of the trait, we encounter errors as below indicating that the future cannot be sent between threads safely.
```
error: future cannot be sent between threads safely
   --> crates/net/dns/src/query.rs:62:51
    |
62  |         self.queued_queries.push_back(Query::Root(Box::pin(resolve_root(resolver, link, timeout))))
    |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ future returned by `resolve_root` is not `Send`
    |
    = help: within `impl Future<Output = Result<(TreeRootEntry, LinkEntry<K>), (LookupError, LinkEntry<K>)>>`, the trait `Send` is not implemented for `impl Future<Output = std::option::Option<std::string::String>>`
note: future is not `Send` as it awaits another future which is not `Send`
```

For more details on this issue, refer to the following resources:
https://blog.rust-lang.org/inside-rust/2023/05/03/stabilizing-async-fn-in-trait.html#mvp-part-2-send-bounds-and-associated-return-types
https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html#async-fn-in-public-traits

On the other hand, by utilizing the `#[async_trait]` attribute, async functions are transformed into methods that return `Pin<Box<dyn Future + Send + 'async_trait>>`. This results in a Future that is thread-safe by default.

### CODE REVIEW

**Comment by AbnerZheng** [comment]
So it is beneficial to introduce `#[trait_variant]` to replace `#[aync_trait]`?

**Comment by AbnerZheng** [comment]
By using `-> impl trait`, we can  avoid using `#[trait_variant]`.  For example we can rewrite the `Resolver` as below:
```rust
pub trait Resolver: Send + Sync + Unpin + 'static {
    /// Performs a textual lookup and returns the first text
    fn lookup_txt(&self, query: &str) -> impl Future<Output = Option<String>> + Send;
}
```
By adding bound `Send` manually, we can avoid the error mentioned above.  And when implementing the trait `Resolver`, we can either use `async fn` or `->impl trait`, both are ok. 
```rust
impl<P: ConnectionProvider> Resolver for AsyncResolver<P> {
    async fn lookup_txt(&self, query: &str) -> Option<String> {
      ...
    }
}
```

```rust
impl<P: ConnectionProvider> Resolver for AsyncResolver<P> {
    fn lookup_txt(&self, query: &str) -> impl Future<Output=Option<String>> + Send {
        todo!()
    }
}
```

**Comment by AbnerZheng** [comment]
WDYT? @mattsse

File: crates/net/dns/src/resolver.rs [hunk 1]
```diff
@@ -1,20 +1,18 @@
...
 use parking_lot::RwLock;
 use std::collections::HashMap;
 use tracing::trace;
 pub use trust_dns_resolver::{error::ResolveError, TokioAsyncResolver};
 use trust_dns_resolver::{name_server::ConnectionProvider, AsyncResolver};
 
 /// A type that can lookup DNS entries
-#[async_trait]
-pub trait Resolver: Send + Sync + Unpin + 'static {
+#[trait_variant::make(Resolver: Send)]
```
Comment (mattsse) [inline crates/net/dns/src/resolver.rs:10]: I don't think we need this

File: crates/net/dns/src/resolver.rs [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
+pub trait LocalResolver: Send + Sync + Unpin + 'static {
     /// Performs a textual lookup and returns the first text
     async fn lookup_txt(&self, query: &str) -> Option<String>;
```
Comment (mattsse) [inline crates/net/dns/src/resolver.rs:13]:
```suggestion
   fn lookup_txt(&self, query: &str) -> impl Future< Output = Option<String>> + Send;
```

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/dns/src/resolver.rs (hunk 2 above)
Comment (AbnerZheng) [reply crates/net/dns/src/resolver.rs]: Done.

File: crates/net/dns/src/resolver.rs (hunk 1 above)
Comment (AbnerZheng) [reply crates/net/dns/src/resolver.rs]: Done.

**Review by mattsse — APPROVED** [review]

**Pull request merged**